_posts/2024-03-20-kraft-migration.md

```diff
@@ -0,0 +1,201 @@
+
+There is a manual procedure, consisting of several phases, to run in order to migrate a cluster without any downtime.
+The following content doesn't want to be an exhaustive description of the migration procedure but more an overview of how it is supposed to work.
+For more details, please refer to the official [ZooKeeper to KRaft Migration](https://kafka.apache.org/documentation/#kraft_zk_migration) documentation.
+
+#### Deploying the KRaft controller quorum
+
+At the beginning, we have the Kafka brokers running in ZooKeeper-mode and connected to the ZooKeeper ensemble used to store metadata.
+
+![ZooKeeper-based cluster](/assets/images/posts/2024-03-20-kraft-migration-01-zk-brokers.png)
+
+> NOTE: the green square boxed number highlights the "generation" of the nodes which are rolled several times during the process.
+
+The first step is to deploy the KRaft controller quorum which will be in charge of maintaining the metadata in KRaft mode.
+In general, the number of KRaft controller nodes will be the same as the number of the ZooKeeper nodes currently running.
+It is also important to highlight that the migration process doesn't support migrating to nodes running in "combined" mode
+The nodes forming the KRaft controller quorum are all configured with the connection to ZooKeeper together with the additional `zookeeper.metadata.migration.enable=true` flag which states the intention to run the migration.
+When the KRaft controllers start, they form a quorum, elect the leader and move in a state where they are waiting for the brokers to register.
+
+![KRaft controller quorum deployed](/assets/images/posts/2024-03-20-kraft-migration-02-kraft-deployed.png)
+
+#### Enabling brokers to run the migration
+
+The next step is to move the brokers to migration mode.
+In order to do so, the brokers' configuration needs to be updated by adding connection details for all nodes of the KRaft controller quorum and enabling the migration with the `zookeeper.metadata.migration.enable=true` flag.
+After the update, the brokers need to be rolled one by one to make such configuration changes effective.
+On restart, the brokers register to the KRaft controller quorum and the migration begins.
+The KRaft controller leader copies all metadata from ZooKeeper to the `__cluster_metadata` topic.
+
+![KRaft migration running](/assets/images/posts/2024-03-20-kraft-migration-03-kraft-migration.png)
+
+While the migration is running, you can verify its status by looking at the log on the KRaft controller leader or by checking the `kafka.controller:type=KafkaController,name=ZkMigrationState` metric.
+When the migration is completed, that is the metric value is `MIGRATION`, the brokers are still running in ZooKeeper mode.
+The KRaft controllers are in charge of handling any requests related to metadata changes within the cluster but they keep sending RPCs to the brokers for metadata updates.
```

Comment:
I'm not complete sure if it will keeps sending RPCs to the brokers for metadata updates at this stage. I think the broker "should" read metadata update from metadata log from KRaft controller, but maybe I'm wrong. Are you sure you're right? Should I start up a cluster to verify it?

Reply:
This is how it works IIRC. Also watching the Kafka Summit presentation about migration was like this. Of course, if you could test to confirm it would be great :-)

Reply:
OK, let me try that later.

Reply:
OK, confirmed you're right. In dual-write mode, the KRaft controller still sends the LeaderAndISR, updateMetadata RPC to brokers.

Here's the log snippet from the KRaft controller while the broker is in the `MIGRATION` state.

```
[2024-03-21 18:06:59,466] INFO [QuorumController id=1] CreateTopics result(s): CreatableTopic(name='quickstart-events2', numPartitions=-1, replicationFactor=-1, assignments=[], configs=[]): SUCCESS (org.apache.kafka.controller.ReplicationControlManager)
[2024-03-21 18:06:59,466] INFO [QuorumController id=1] Replayed TopicRecord for topic quickstart-events2 with topic ID btjye7G1Q9OsSh6xd-xXgQ. (org.apache.kafka.controller.ReplicationControlManager)
[2024-03-21 18:06:59,466] INFO [QuorumController id=1] Replayed PartitionRecord for new partition quickstart-events2-0 with topic ID btjye7G1Q9OsSh6xd-xXgQ and PartitionRegistration(replicas=[0], directories=[gYNd_gPZql98DYeZ-5NAQw], isr=[0], removingReplicas=[], addingReplicas=[], elr=[], lastKnownElr=[], leader=0, leaderRecoveryState=RECOVERED, leaderEpoch=0, partitionEpoch=0). (org.apache.kafka.controller.ReplicationControlManager)
...
[2024-03-21 18:06:59,518] INFO [Controller id=1 epoch=3] Sending UpdateMetadata request to brokers HashSet() for 0 partitions (state.change.logger)
[2024-03-21 18:06:59,518] INFO [Controller id=1 epoch=3] Sending LeaderAndIsr request to broker 0 with 1 become-leader and 0 become-follower partitions (state.change.logger)
[2024-03-21 18:06:59,518] INFO [Controller id=1 epoch=3] Sending UpdateMetadata request to brokers HashSet(0) for 1 partitions (state.change.logger)
```

Reply:
Yeah, the active KRaft controller first sends full RPCs and then switch to incremental mode, until migration is finalized.

Reply:
Thanks for confirmation @showuon ;-)